### PR TITLE
TVRN-13: SQL Init Script

### PR DESCRIPTION
**Pull request by Tauntalus**
Pull request, part 2! The initialization script is finished, and encompasses all the tables we hashed out last quarter.

The idempotency is not yet implemented - currently, there's no data insertion or shuffling, only the creation of new tables, which will throw an error if the tables already exist. This can be cleaned up, but I want to get the starting point published before going forward.

### CODE REVIEW

**Comment by Shadow53** [comment]
> The idempotency is not yet implemented - currently, there's no data insertion or shuffling, only the creation of new tables, which will throw an error if the tables already exist. This can be cleaned up, but I want to get the starting point published before going forward.

With an idempotent script, we don't want it to error if it is run on an already set-up database. It should run without errors and just not do anything (i.e. succeed).

File: tavern_server/src/db_tables.sql [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+	'Tiny', 
+	'Small', 
+	'Medium', 
+	'Large', 
+	'Huge', 
+	'Gargantuan', 
+	'Colossal'
+);
+
+CREATE TYPE Alignment AS ENUM (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Thought: do we want to have alignment set up like this or as two fields: good/evil and lawful/chaotic?

File: tavern_server/src/db_tables.sql [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+	'CMB',
+	'CMD',
+	'Armor Class',
+	'Touch AC',
+	'Flat-Footed AC',
+	'Initiative Bonus',
+	'Damage Reduction',
+	'Spell Resistance',
+	'Speed',
+	'Fortititude Save',
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Typo: one too many "ti" in "Fortitude"

File: tavern_server/src/db_tables.sql [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+	'Reflex Save',
+	'Will Save'
+);
+
+CREATE TYPE RaceType AS ENUM (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: What about racial subtypes, e.g. the many subtypes of humanoid?

File: tavern_server/src/db_tables.sql [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+	'Energy', 
+	'Acid', 
+	'Fire', 
+	'Electricity', 
+	'Cold', 
+	'Sonic', 
+	'Positive', 
+	'Negative', 
+	'Nonlethal'
+);
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Maybe these should be separated into multiple categories? A sword enchanted with fire could do slashing and fire damage, right?

File: tavern_server/src/db_tables.sql [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+
+CREATE TYPE Material AS ENUM (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: I'm not sure this should be an enum. More likely this would be better as a table where we use foreign keys to indicate something's material.

File: tavern_server/src/db_tables.sql [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+        salt        		BYTEA,
+
+        time_cost   		INT,
+        memory      		INT,
+        threads     		INT
+);
+
+--Character overview table
+CREATE TABLE Character (
+        user_id         	UUID,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Should be a foreign key

File: tavern_server/src/db_tables.sql [hunk 7]
```diff
@@ -0,0 +1,676 @@
...
+
+	levels_taken		INT,
+	hp_taken		INT,
+	skills_taken		INT
+);
+
+--Feats tables
+CREATE TABLE Feats (
+	feat_id			UUID,
+	req_id			UUID,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: What if a feat has multiple requirements?

File: tavern_server/src/db_tables.sql [hunk 8]
```diff
@@ -0,0 +1,676 @@
...
+	attr_unit_id		UUID,
+	req_attr		Attribute,
+	score			INT
+);
+
+--this table creates a many to many realationship internally between feats.
+--any given feat can require a number of other feats to be taken,
+--and a given feat can be required by any number of other feats.
+--this table accomplishes that.
+CREATE TABLE RequiredFeats (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: I think `FeatRequirements` is a more descriptive name for this

File: tavern_server/src/db_tables.sql [hunk 9]
```diff
@@ -0,0 +1,676 @@
...
+	class_id		UUID,
+	item_id			UUID
+);
+
+CREATE TABLE ClassProficientWeaponClass (
+	class_id		UUID,
+	weapon_class		WeaponClass
+);
+
+CREATE TABLE ClassProficientWeaponAllBut (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: I think we discussed this before: perhaps it would be better to indicate weapons that the class specifically is not proficient in? Then that could be combined with the weapon type proficiencies to create the same effect while being more flexible

File: tavern_server/src/db_tables.sql [hunk 10]
```diff
@@ -0,0 +1,676 @@
...
+	 class_id		UUID,
+	weapon_class		WeaponClass
+);
+
+CREATE TABLE ClassProficientArmorClass (
+	class_id		UUID,
+	armor_class		ArmorClass
+);
+
+CREATE TABLE ClassProficientArmorAllBut (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: See comment on equivalent weapon table

File: tavern_server/src/db_tables.sql [hunk 11]
```diff
@@ -0,0 +1,676 @@
...
+
+CREATE TABLE DomainSpells (
+	domain_id		UUID,
+	spell_id		UUID,
+	casts			INT
+);
+
+CREATE TABLE SpellComponents (
+	spell_id		UUID,
+	item_id			UUID,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: This looks like it takes into account only material components. What about voice and hand-waving? (Forgetting the correct terms for those atm)

**Review by Shadow53 — CHANGES_REQUESTED** [review]
I've left a handful of notes throughout the file changes. Please take a look at them.
Something I noticed throughout the file is that nothing was marked as primary key, as foreign key, or with any sort of consistency checks (e.g. value > 0). Please include those as well before this can be approved.

File: tavern_server/src/db_tables.sql (hunk 1 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: The Pathfinder SRD has moments where it's easier to treat alignments as two separate axes, and moments where treating them as 9 elements of the same ilk is easier. It really depends on which problems we'd rather deal with - splitting one element into two, or remerging two disparate elements into one. I lean towards the former.

File: tavern_server/src/db_tables.sql [hunk 12]
```diff
@@ -0,0 +1,676 @@
...
+	'Sunsilver',
+	'Viridium',
+	'Voidglass',
+	'Whipwood',
+	'Wyroot'
+);
+
+--table declarations
+--user creds table - also used for hashing info
+CREATE TABLE UserInfo (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: BTW, I would prefer if this were called `Users` rather than `UserInfo`. When I took a database class, the general naming rule we went by was that a table is named the plural of the thing stored in it, or a combination of two things for junction tables.

File: tavern_server/src/db_tables.sql [hunk 13]
```diff
@@ -0,0 +1,676 @@
...
+	user_id     		UUID,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Might as well follow a convention of "this" thing's id being `id`, and all foreign keys are `thing_id`.

File: tavern_server/src/db_tables.sql [hunk 14]
```diff
@@ -0,0 +1,676 @@
...
+
+        name        		TEXT,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Is this user's name ("Michael") or username ("Shadow53")?
We definitely need either a username or an email for the purposes of logging in. Honestly, it would probably be better to use an email address for that.

File: tavern_server/src/db_tables.sql (hunk 7 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: fixed; requirements have been changed to reference the feat_id primary key, allowing a feat to have multiple requirements. Each of the ___Requirement tables have been updated to use feat_id as well. I don't remember why I had this foreign key setup, admittedly.

File: tavern_server/src/db_tables.sql (hunk 6 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: noted.

File: tavern_server/src/db_tables.sql (hunk 5 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Noted, I can implement this before the pull request goes out.

File: tavern_server/src/db_tables.sql (hunk 3 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: I noticed the subtypes, but chose to omit them at first. I can add a RaceSubtype ENUM for the various racial subtypes - admittedly, I assumed they wouldn't be very important, but the SRD shows that subtypes matter a bit more than merely flavor and cosmetic.

File: tavern_server/src/db_tables.sql (hunk 2 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Whoops! Fixed, thanks for pointing that out. Mildly embarasasasasassing.

File: tavern_server/src/db_tables.sql (hunk 8 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Implemented. I feel RequiredFeats is a better descriptor, since FeatRequirements makes the table sound like it lists the requirements *for* the feat, not the feats *required* by it. However, I respect the appeal to consistency in nomenclature, so it's been updated.

File: tavern_server/src/db_tables.sql (hunk 9 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: The suite of tables here was meant to accomplish this, but without arrays. Now that I know that PostgreSQL supports columns with array types, this has been cleaned up as you suggested. I feel like it's more intuitive for the table to contain the weapons a class *is* proficient in, though. Implemented all the same.

File: tavern_server/src/db_tables.sql (hunk 11 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Implemented with a new field and ENUM called component_type and SpellComponent, respectively.

File: tavern_server/src/db_tables.sql (hunk 12 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Implemented.

File: tavern_server/src/db_tables.sql (hunk 13 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Implemented.

File: tavern_server/src/db_tables.sql (hunk 14 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Thanks for clearing this up for me, this is definitely design cruft from when I first wrote down the DB ERD. I've changed name to "username" and added an email field. Added a nickname field as well for easy implementation of user profile management options later on.

File: tavern_server/src/db_tables.sql (hunk 4 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: That could work. I've changed the damage type column on a weapon into an array, so a sword can have multiple types of damage. Also worth noting, in the table Weapons, damage has been made an array of TEXT items (TEXTis a placeholder for dice shorthand, IE 2d6)

**Comment by Tauntalus** [comment]
Just finished fleshing the database out with constraints, refactored tables, and labeled primary keys and foreign keys. I hit every point raised in the above thread, and did my best to cover my bases this time around. I don't believe I've missed anything, but hey - that's what review is for.

File: tavern_server/src/db_tables.sql [hunk 15]
```diff
@@ -0,0 +1,861 @@
...
+	'Monstrous Humanoid', 
+	'Ooze', 
+	'Outsider', 
+	'Plant', 
+	'Undead', 
+	'Vermin'
+);
+
+CREATE TYPE IF NOT EXISTS RaceSubtyep AS ENUM (
+
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: You seem to have left this enum empty...

File: tavern_server/src/db_tables.sql [hunk 16]
```diff
@@ -0,0 +1,861 @@
...
+	'Cold', 
+	'Sonic', 
+	'Positive', 
+	'Negative', 
+	'Nonlethal'
+);
+
+--table declarations
+--user creds table - also used for hashing info
+CREATE TABLE IF NOT EXISTS Users (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:208]: Formatting looks off here.

File: tavern_server/src/db_tables.sql [hunk 17]
```diff
@@ -0,0 +1,861 @@
...
+	id     			UUID PRIMARY KEY,
+
+	email			TEXT 
+		NOT NULL,
+        username       		TEXT 
+		NOT NULL,
+	nickname		TEXT
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Do we need this? I imagine we could just display the username. If we keep it, I think it can be null - we would use the username as default.

File: tavern_server/src/db_tables.sql [hunk 18]
```diff
@@ -0,0 +1,861 @@
...
+		NOT NULL,
+        salt        		BYTEA 
+		NOT NULL,
+
+        time_cost   		INT 
+		NOT NULL,
+        memory      		INT 
+		NOT NULL,
+        threads     		INT 
+		NOT NULL
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Add a `CHECK` constraint to these to make sure they are positive.

File: tavern_server/src/db_tables.sql [hunk 19]
```diff
@@ -0,0 +1,861 @@
...
+	'Negative', 
+	'Nonlethal'
+);
+
+--table declarations
+--user creds table - also used for hashing info
+CREATE TABLE IF NOT EXISTS Users (
+	id     			UUID PRIMARY KEY,
+
+	email			TEXT
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Should we include a check here to ensure the email address fits a pattern? It seems possible in Postgres.

File: tavern_server/src/db_tables.sql [hunk 20]
```diff
@@ -0,0 +1,861 @@
...
+        time_cost   		INT 
+		NOT NULL,
+        memory      		INT 
+		NOT NULL,
+        threads     		INT 
+		NOT NULL
+);
+
+--Character overview table
+CREATE TABLE IF NOT EXISTS Characters (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:237]: Again, formatting seems off.

File: tavern_server/src/db_tables.sql [hunk 21]
```diff
@@ -0,0 +1,861 @@
...
+        char_id         	UUID PRIMARY KEY,
+
+	id			UUID REFERENCES Users 
+		NOT NULL,
+        race_id         	UUID REFERENCES Races 
+		NOT NULL,
+        deity_id        	UUID REFERENCES Deities 
+		NOT NULL,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:245]: I believe you need to specify which column each foreign key references, not just the table.

File: tavern_server/src/db_tables.sql [hunk 22]
```diff
@@ -0,0 +1,861 @@
...
+        age             	INT 
+		NOT NULL 
+		CHECK (age > 0),
+        gender          	Gender,
+        alignment       	Alignment 
+		NOT NULL,
+        backstory       	TEXT,
+        height          	INT,
+        weight          	INT,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Should be `> 0`

File: tavern_server/src/db_tables.sql [hunk 23]
```diff
@@ -0,0 +1,861 @@
...
+		CHECK (constitution >= 0 AND constitution <= 20),
+        intelligence    	INT 
+		NOT NULL
+		CHECK (intelligence >= 0 AND intelligence <= 20),
+        wisdom          	INT
+		NOT NULL
+		CHECK (wisdom >= 0 AND wisdom <= 20),
+        charisma        	INT
+		NOT NULL
+		CHECK (charisma >= 0 AND charisma <= 20),
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:281]: Are you sure stats can *never* be greater than 20?

File: tavern_server/src/db_tables.sql [hunk 24]
```diff
@@ -0,0 +1,861 @@
...
+--Race tables
+CREATE TABLE IF NOT EXISTS Races (
+	race_id 		UUID PRIMARY KEY,
+	type_id			UUID REFERENCES RaceTypes
+		NOT NULL,
+	subtype_id		UUID REFERENCES RaceSubtypes,
+
+	name			TEXT
+		NOT NULL,
+	move_speed		INT
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:316]: `> 0`

File: tavern_server/src/db_tables.sql [hunk 25]
```diff
@@ -0,0 +1,861 @@
...
+		NOT NULL,
+	size			Size
+		NOT NULL,
+	languages		TEXT
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: If we have a place where languages are defined, this should either be an array of those (I know nothing about that feature) or this should be tied to a junction table so we can easily determine how many languages are available to the race and which ones.

File: tavern_server/src/db_tables.sql [hunk 26]
```diff
@@ -0,0 +1,861 @@
...
+
+	name			TEXT
+		NOT NULL,
+	hit_die			TEXT
+		NOT NULL,
+	bab_per_hit_die		REAL
+		NOT NULL
+);
+
+CREATE TABLE IF NOT EXiSTS RaceSubtypes (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Needs a column to keep track of the supertype. Also make sure you can't have two+ supertypes.

File: tavern_server/src/db_tables.sql [hunk 27]
```diff
@@ -0,0 +1,861 @@
...
+	
+--Class tables
+CREATE TABLE IF NOT EXISTS Classes (
+	class_id		UUID PRIMARY KEY,
+
+	name			TEXT
+		NOT NULL,
+	hit_die			TEXT
+		NOT NULL,
+	starting_wealth		TEXT
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:363]: `>= 0`?

File: tavern_server/src/db_tables.sql [hunk 28]
```diff
@@ -0,0 +1,861 @@
...
+
+CREATE TABLE IF NOT EXISTS characterSubclasses (
+	char_id			UUID REFERENCES Characters
+		NOT NULL,
+	subclass_id		UUID REFERENCES Subclasses
+		NOT NULL,
+
+	levels_taken		INT
+		NOT NULL
+		CHECK (levels_taken > 0),
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:377]: We should discuss if it is possible to have a level 0 character, e.g. one saved in the middle of creation.

File: tavern_server/src/db_tables.sql [hunk 29]
```diff
@@ -0,0 +1,861 @@
...
+	skills_taken		INT
+		NOT NULL
+		CHECK (skills_taken >= 0)
+);
+
+--Feats tables
+CREATE TABLE IF NOT EXISTS Feats (
+	feat_id			UUID PRIMARY KEY,
+
+	short_description	TEXT,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Why is this nullable?

File: tavern_server/src/db_tables.sql [hunk 30]
```diff
@@ -0,0 +1,861 @@
...
+	description		TEXT
+		NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS ClassFeatures (
+	class_id		UUID REFERENCES Feats
+		NOT NULL,
+	feature_id		UUID REFERENCES Features
+		NOT NULL,
+	is_default		BOOLEAN
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:469]: For this and other `BOOLEAN` columns, you should define a default value to each. You may leave `NOT NULL` if a column with a default value can still somehow be `NULL`.

File: tavern_server/src/db_tables.sql [hunk 31]
```diff
@@ -0,0 +1,861 @@
...
+	casts			INT
+		NOT NULL
+		CHECK (casts >= 0)
+);
+
+CREATE TABLE IF NOT EXISTS SubclassSpells (
+	subclass_id		UUID REFERENCES Subclasses,
+	spell_id		UUID REFERENCES Spells,
+	casts			INT,
+	req_level		INT
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Seems like these should be marked `NOT NULL`

File: tavern_server/src/db_tables.sql [hunk 32]
```diff
@@ -0,0 +1,861 @@
...
+);
+
+CREATE TABLE IF NOT EXISTS DomainSpells (
+	domain_id		UUID REFERENCES Domains,
+	spell_id		UUID REFERENCES Spells,
+	casts			INT
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:554]: Same for this: why not `NOT NULL`?

File: tavern_server/src/db_tables.sql [hunk 33]
```diff
@@ -0,0 +1,861 @@
...
+);
+
+CREATE TABLE IF NOT EXISTS SpellComponents (
+	spell_id		UUID REFERENCES Spells
+		NOT NULL,
+	item_id			UUID REFERENCES Items,
+
+	amount			INT,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: What does `amount` mean in this context? Is it not required?

File: tavern_server/src/db_tables.sql [hunk 34]
```diff
@@ -0,0 +1,861 @@
...
+		NOT NULL,
+	favored_animals		TEXT
+);
+
+CREATE TABLE IF NOT EXISTS DeityDomains (
+	deity_id		UUID REFERENCES Deities
+		NOT NULL,
+	domain_id		UUID REFERENCES Domains
+		NOT NULL 	--no deities without domains
+);
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:608]: Is there a way to require that every deity has at least one domain? Some sort of `CHECK` or other constraint, perhaps?

File: tavern_server/src/db_tables.sql [hunk 35]
```diff
@@ -0,0 +1,861 @@
...
+
+CREATE TABLE IF NOT EXISTS DeityWeapons (
+	deity_id		UUID REFERENCES Deities
+		NOT NULL,
+	item_id			UUID REFERENCES Items
+);
+
+--Item tables
+CREATE TABLE IF NOT EXISTS PathfinderItems (
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:617]: Why `PathfinderItems` and not just `Items`?

File: tavern_server/src/db_tables.sql [hunk 36]
```diff
@@ -0,0 +1,861 @@
...
+	item_id			UUID PRIMARY KEY,
+
+	name			TEXT
+		NOT NULL,
+	description		TEXT,
+	cost			INT	--prices are in copper
+		NOT NULL
+		CHECK (cost >= 0),
+	weight			INT
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql:626]: Can something have negative weight?

**Review by Shadow53 — CHANGES_REQUESTED** [review]
I've left some more comments. Main things are:
- Some things are left nullable that I'm not sure should be
- Formatting is a little weird on some lines - I'd like to see one column definition per line, if possible
- Double-check that you aren't missing any possible constraints on columns. If something must hold true about the data, I would like to see it reflected here if possible

File: tavern_server/src/db_tables.sql (hunk 23 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:281]: The stored stat can never be greater than 20, but the _computed value_ can be greater than 20. If your Strength is at 20, you can't put another level into it - but you can drink a Potion of Strength for a +2 to the stat.

File: tavern_server/src/db_tables.sql (hunk 29 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Short descriptions were nullable because, by default, everything has a "long description." Short descriptions, I presumed, would be saved for incredibly common feats that are almost always taken. The same applies to spells and class features.

File: tavern_server/src/db_tables.sql (hunk 35 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:617]: When I attempted to test and compile the script during early testing, I found that "Items" was an existing relation. (table) I have been using PathfinderItems instead to circumvent this, although it may have been a quirk of the checker I was using.

File: tavern_server/src/db_tables.sql (hunk 34 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:608]: Not that I know of, sadly. I can look into this as an additional constraint to implement, but there doesn't seem to be a built-in way to ensure that every entry in Deities links to at least one in the Domains table.

File: tavern_server/src/db_tables.sql (hunk 33 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: renamed to item_amount - it is not required if the component is Verbal or Somatic.

File: tavern_server/src/db_tables.sql (hunk 17 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Fair. I'll remove the nickname field - if we need it, we can add it in later.

File: tavern_server/src/db_tables.sql (hunk 15 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Turns out this was a memory leak - a leak from my memory. I ended up putting both RaceType and RaceSubtype into tables instead of enums (similar to the handling of materials) and have removed both of these since.

File: tavern_server/src/db_tables.sql (hunk 18 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: noted. obvious in hindsight, but I wasn't sure.

File: tavern_server/src/db_tables.sql (hunk 19 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: I can look into this , but also "email regex" sends shivers down my spine.

File: tavern_server/src/db_tables.sql (hunk 21 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:245]: This is actually something I found out - if you only specify the table, not the foreign key, then Postgres immediately assumes it to be the primary key of the referenced table. Easy shorthand that sped up the writing process.

File: tavern_server/src/db_tables.sql (hunk 22 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Implemented.

File: tavern_server/src/db_tables.sql (hunk 24 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:316]: Implemented.

File: tavern_server/src/db_tables.sql (hunk 25 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Implemented as an array. They work like you'd expect, being an array of values. I'm not sure how joins work on this, but I also am not designing it with that in mind. As far as I'm designing the database, array types are only for data endpoints, never keys.

File: tavern_server/src/db_tables.sql (hunk 19 above)
Comment (Shadow53) [reply tavern_server/src/db_tables.sql]: As it [well should](https://www.regular-expressions.info/email.html). I was thinking, if we did any checking, it would be a general `[thing]@[thing].[thing]`, so `a@b.c` would be a valid email address. Any extra validation would come later in the form of sending a confirmation email before the account is fully created.

File: tavern_server/src/db_tables.sql (hunk 30 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:469]: Implemented.

File: tavern_server/src/db_tables.sql (hunk 25 above)
Comment (Shadow53) [reply tavern_server/src/db_tables.sql]: >  I'm not sure how joins work on this, but I also am not designing it with that in mind.

I think I would prefer a junction table, then, because we know that we can do `SELECT`  queries and `JOIN`s using those. If you can find that it's just as easy to do with arrays, I'm fine with keeping it, though.

File: tavern_server/src/db_tables.sql (hunk 16 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:208]: I'm not sure what you mean - here on GitHub it looks fine, and it looks fine in my terminal. I'll give a quick reformat to try and fix any issues that just aren't showing up on my box, though.

File: tavern_server/src/db_tables.sql (hunk 29 above)
Comment (Shadow53) [reply tavern_server/src/db_tables.sql]: Arguably, if anything, the long description should be the nullable one. Consider from a UI perspective: is the user more likely to see a short description (e.g. in a list of things) or the long one (i.e. in the item description)? Further, the short description can be used in the place of the long one without issue, but the long one is too long to go in place of the short one -- if that phrasing makes sense.

TL;DR: requiring short descriptions is better for UX.

File: tavern_server/src/db_tables.sql (hunk 35 above)
Comment (Shadow53) [reply tavern_server/src/db_tables.sql:617]: Huh, weird. Which checker were you using? Have you tried running the script with `Items` against an actual database?

File: tavern_server/src/db_tables.sql (hunk 34 above)
Comment (Shadow53) [reply tavern_server/src/db_tables.sql:608]: Might be able to do something using triggers, but I'd have to look into that before I could say one way or another. Let's drop this for now and we can look at making it enforced later.

File: tavern_server/src/db_tables.sql (hunk 16 above)
Comment (Shadow53) [reply tavern_server/src/db_tables.sql:208]: There were some tables with columns where the `NOT NULL` appears to be on a separate line from the column it modifies, and oddly indented.

File: tavern_server/src/db_tables.sql (hunk 32 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:554]: Because I forgot. Implemented!

File: tavern_server/src/db_tables.sql (hunk 31 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Implemented.

File: tavern_server/src/db_tables.sql (hunk 28 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql:377]: I agree. I know a lot of people do it piecemeal, but it does mean letting a lot of columns be nullable, or link to default "empty" classes and races and spells. I don't think that's ideal, admittedly, but I'm open to being persuaded.

File: tavern_server/src/db_tables.sql (hunk 26 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: Admittedly I don't think we do - I couldn't find a playable race with more than one Creature Type, or more than one Creature Subtype. I've made races many-to-one with both Types and Subtypes as a result - the insurance of "one subtype and one supertype" is handled in the Races table by...only having a singular UUID for subtype and supertype respectively.

File: tavern_server/src/db_tables.sql [hunk 37]
```diff
@@ -0,0 +1,860 @@
...
+		NOT NULL,
+        deity_id        	UUID REFERENCES Deities 
+		NOT NULL,
+
+        name            	TEXT 
+		NOT NULL,
+        age             	INT 
+		NOT NULL 
+		CHECK (age > 0),
+        gender          	Gender,
```
Comment (Shadow53) [inline tavern_server/src/db_tables.sql]: Seems weird that gender would be nullable, of all things

File: tavern_server/src/db_tables.sql (hunk 29 above)
Comment (Shadow53) [reply tavern_server/src/db_tables.sql]: I still stand by this, so I would appreciate if you could change it.

**Review by Shadow53 — CHANGES_REQUESTED** [review]
I would still like to see short description be the required one and long description be nullable. I also made a note about `gender` being nullable, which seems weird. That one is not a huge deal, though.

It doesn't seem like you have implemented any tests for this. I would suggest adding the necessary Rust code to run this against a configured database and then write the integration tests before we merge this. My concern is that an error would occur if this were actually run against a database, and I would like to see the tests showing this is written correctly.

That said, I think this is close enough to finished that we can work on other things as well and assume that the columns and their names will not change.

File: tavern_server/src/db_tables.sql (hunk 37 above)
Comment (Tauntalus) [reply tavern_server/src/db_tables.sql]: I ran with the idea that players would focus on their character mechanically first, aesthetically next. Everything mechanical should be completely mandatory, and anything aesthetic or nonfunctional can be left to the user's discretion. Still, you raise a valid point - I don't think anybody builds a character without thinking of their gender as one of the starting points.

TLDR fixing and implementing.

**Comment by Shadow53** [comment]
Looks good to me so far. There are probably further constraints we can add if we dissect it even further, but we can worry about that later.
I would like to see one or two basic tests that run this against an actual database, to make sure there aren't any errors that occur. You'll want to do this following the suggestions in the Rust Book for [integration tests](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests). As I've mentioned before, I think this is close enough to done that we can use it as a base for writing other DB code. Your work on getting the `tavern_pathfinder` library started is more important at this point, but once the foundation of that work is finished and merged, please come back to this and write the test(s). Then I'll go ahead and approve this for merging.

File: tavern_server/src/db.rs [hunk 38]
```diff
@@ -23,6 +23,10 @@ mod tests {
         assert!(conn_string.contains("user=foo"));
         assert!(conn_string.contains("password=bar"));
     }
+    #[test]
+    fn postgres_database_init() {
+        assert!(true);
```
Comment (Shadow53) [inline tavern_server/src/db.rs:28]: As future reference, it's better to make an unimplemented test fail than succeed. Also, the `unimplemented!()` macro exists for this.

**Review by Shadow53 — APPROVED** [review]
